renderer/execute/types.py

```diff
@@ -128,31 +130,38 @@ def _as_i18n_message(self) -> I18nMessage:
-                # 1 other -- if there were 1 other, we might as well have
-                # written the name itself)
-                names[2:] = [f"{len(names) - 2} others"]
-            if len(names) == 1:
-                columns_str = f"The column {names[0]}"
-            else:
-                # English-style:
-                # 2: "A" and "B"
-                # 3: "A", "B" and "C"
-                # 4+: "A", "B" and 2 others
-                names_str = ", ".join(names[:-1]) + " and " + names[-1]
-                columns_str = f"The columns {names_str}"
+            icu_args = {
+                "columns": len(self.column_names),
+                **{
+                    str(i): self.column_names[i]
+                    for i in range(0, len(self.column_names))
+                },
+            }
 
             if self.should_be_text:
                 # Convert to Text
-                return I18nMessage.TODO_i18n(
-                    f"{columns_str} must be converted to Text."
+                # i18n: The parameter {columns} will contain the total number of columns that need to be converted; you will also receive the column names as {0}, {1}, {2}, etc.
+                return I18nMessage.trans(
+                    "py.renderer.execute.types.PromptingError.WrongColumnType.as_error_message.shouldBeText",
+                    default="{ columns, plural, offset:2"
+                    " =1 {The column “{0}” must be converted to Text.}"
+                    " =2 {The columns “{0}” and “{1}” must be converted to Text.}"
+                    " one {The columns “{0}”, “{1}” and “{2}” must be converted to Text.}"
+                    " other {The columns “{0}”, “{1}” and # others must be converted to Text.}}",
+                    args=icu_args,
                 )
             else:
-                return I18nMessage.TODO_i18n(
-                    f"{columns_str} must be converted "
-                    f"from {self.found_type_name} to {self.best_wanted_type_name}."
+                icu_args["found_type"] = self.found_type
+                icu_args["best_wanted_type"] = self.best_wanted_type_id
+                # i18n: The parameter {columns} will contain the total number of columns that need to be converted; you will also receive the column names: {0}, {1}, {2}, etc. The parameters {found_type} and {best_wanted_type} will have values among "text", "number", "datetime"; however, including a (possibly empty) "other" case is mandatory.
+                return I18nMessage.trans(
+                    "py.renderer.execute.types.PromptingError.WrongColumnType.as_error_message.general",
+                    default="{ columns, plural, offset:2"
+                    " =1 {The column “{0}” must be converted from { found_type, select, text {Text} number {Numbers} datetime {Dates & Times} other {}} to {best_wanted_type, select, text {Text} number {Numbers} datetime {Dates & Times} other {}}.}"
```

Comment:
We plan to add new types to Workbench. And I've _almost_ convinced Pierre that we should rename "Dates & Times" to "Timestamps" (making way for a new "Date" type). I worry it will be costly to maintain these `select` clauses in all pofiles.

Should we translate the types separately?

Reply:
In general, it's difficult (or impossible) for the result to make (ideal) sense in any possible (human) language when we translate it in parts and combine them with code. Even if we assume that the syntax will make sense (as we are tempted -and, who knows, maybe entitled?- to do in this case), some languages are heavily inflected (i.e. may need different forms for some terms depending on their syntactic function); for example, a simple Greek translation of the message you commented on would require the `best_wanted_type` translations to be in accusative, but requiring a translation in accusative might conflict with other languages.

Another reason that we can't compose messages is the current structure of `I18nMessage`s, although we could overcome this. Currently, the parameters of an `I18nMessage` can only be plain data (number or string); if we want to compose messages, we must allow the type `I18nMessage` for parameters of `I18nMessage`s.

Reply:
We don't inflect column names; and I don't think users expect us to. Could some creative phrasing, punctuation or formatting convey "this word is data, not a normal word?" Quotation marks / small-caps / underline / ...?

As for the structure: you make a good point -- there's a solid challenge here. I can imagine customizing ICU to let Greek translators write `{found_type, cjwcolumntype, accusative}`. ICU does not [encourage](http://userguide.icu-project.org/formatparse/messages#TOC-Custom-Format-Objects-discouraged-) this, but it's probably doable. This would involve some careful code per language, once.

So we're on the same page, could you please describe to me what we'd need to do to add a new type to Workbench using the strategy of this pull request as-is? Keep in mind that modules' messages may include data-type names, too. And when the number of Workbench modules explodes, we on the core team won't be able to revise modules that aren't ours.

My thesis: we're committing to a _ton_ of effort here, regardless of our choice.

Reply:
I wasn't referring to inflected column names, but to inflected column types. If we translate data types separately, some trick such as the `column "A"` we use for the column name could probably be used by the translators for the column types (I don't know if it would sound ideal in _all_ languages, but it would probably work for all languages in some way). But, of course, we would need to make `I18nMessage` able to store `I18nMessage` as a parameter.

With the current status of the pull request, in order to add a new data type, we would indeed need to add a new case to all the `select` messages (anywhere the po files are stored) that have data types as cases.

`{found_type, cjwcolumntype, accusative}` looks kind of weird

Reply:
Thank you for talking all this through with me. I'm inclined to table my concern for now. We can redesign the types approach later, if we find the `select` clauses are too costly in practice.